core/src/main/java/org/infinispan/marshall/core/internal/InternalMarshaller.java

```diff
@@ -0,0 +1,227 @@
+      BytesObjectOutput out = writeObjectOutput(obj);
+      return out.toBytes(); // trim out unused bytes
+   }
+
+   private BytesObjectOutput writeObjectOutput(Object obj) throws IOException {
+      BufferSizePredictor sizePredictor = marshallableTypeHints.getBufferSizePredictor(obj);
+      BytesObjectOutput out = writeObjectOutput(obj, sizePredictor.nextSize(obj));
+      sizePredictor.recordSize(out.pos);
+      return out;
+   }
+
+   private BytesObjectOutput writeObjectOutput(Object obj, int estimatedSize) throws IOException {
+      BytesObjectOutput out = new BytesObjectOutput(estimatedSize, this);
+      Externalizer<Object> ext = externalizers.findWriteExternalizer(obj, out);
+      if (ext != null)
+         ext.writeObject(out, obj);
+      else
+         external.objectToObjectStream(obj, out);
+
+      return out;
+   }
+
+   @Override
+   public Object objectFromByteBuffer(byte[] buf) throws IOException, ClassNotFoundException {
+      ObjectInput in = BytesObjectInput.from(buf, this);
+      return objectFromObjectInput(in);
+   }
+
+   private Object objectFromObjectInput(ObjectInput in) throws IOException, ClassNotFoundException {
+      Externalizer<Object> ext = externalizers.findReadExternalizer(in);
+      if (ext != null)
+         return ext.readObject(in);
+      else {
+         try {
+            return external.objectFromObjectStream(in);
+         } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return null;
+         }
+      }
+   }
+
+   @Override
+   public ObjectOutput startObjectOutput(OutputStream os, boolean isReentrant, int estimatedSize) throws IOException {
+      BytesObjectOutput out = new BytesObjectOutput(estimatedSize, this);
+      return new StreamBytesObjectOutput(os, out);
+   }
+
+   @Override
+   public void objectToObjectStream(Object obj, ObjectOutput out) throws IOException {
+      out.writeObject(obj);
+   }
+
+   @Override
+   public void finishObjectOutput(ObjectOutput oo) {
+      try {
+         oo.flush();
+      } catch (IOException e) {
+         // ignored
+      }
+   }
+
+   @Override
+   public Object objectFromByteBuffer(byte[] buf, int offset, int length) throws IOException, ClassNotFoundException {
+      // Ignore length since boundary checks are not so useful here where the
+      // unmarshalling code knows what to expect specifically. E.g. if reading
+      // a byte[] subset within it, it's always appended with length.
+      BytesObjectInput in = BytesObjectInput.from(buf, offset, this);
+      return objectFromObjectInput(in);
+   }
+
+   @Override
+   public Object objectFromInputStream(InputStream is) throws IOException, ClassNotFoundException {
+      // This is a very limited use case, e.g. reading from a JDBC ResultSet InputStream
+      // So, this copying of the stream into a byte[] has not been problematic so far,
+      // though it's not really ideal.
```

Comment:
I think you could rework the `PositionalBuffer` interfaces to be more friendly with `InputStream`'s `mark`/`reset` idiom, and you wouldn't need to copy it into a `byte[]`.